src/journal/journald-context.c

```diff
@@ -268,6 +274,54 @@ static int client_context_read_label(
         return 0;
 }
 
+static int client_context_read_log_filter_xattr(const char *cgroup, Set **patterns, const char *key) {
+        _cleanup_free_ char *xattr = NULL;
+        int r;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, cgroup, key, &xattr);
+        if (r == -ENODATA) {
+                /* -ENODATA is returned when the key is not defined, this is not an error. */
+                *patterns = set_free_with_destructor(*patterns, pattern_free);
+                return 0;
+        }
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get %s xattr for %s: %m", key, cgroup);
+
+        const char *pattern;
+        NULSTR_FOREACH(pattern, xattr) {
+                pcre2_code *regex = NULL;
+
+                r = pattern_compile_and_log(pattern, 0, &regex);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to compile log filtering pattern '%s' for %s: %m",
+                                               pattern, cgroup);
+
+                r = set_ensure_put(patterns, NULL, regex);
+                if (r < 0) {
+                        regex = pattern_free(regex);
+                        return log_debug_errno(r, "Failed to insert regex into set for %s: %m", cgroup);
+                }
```

Comment:
hmm, so this is purely additive right now. You keep adding new entries, but if an entry vanished from the xattr you never remove it. Hence I really would always start fresh and allocate a new Set object, and add to it, and then return that, so that the list is always comprehensive, and correct.